### ISSUE TITLE
Produce Transformers.js v4 ONNX conversions for TinyBERT and MobileBERT

### ISSUE THREAD

**Issue by atjsh** [atjsh/llmlingua-2-js#28]
## Summary

Produce neutral, reproducible Transformers.js v4 ONNX format-conversion archives for the existing JueZhang TinyBERT and MobileBERT LLMLingua-2 checkpoints.

Publication trusts the pinned official converter command: a successfully emitted artifact is uploaded even when validation warns, a runtime cannot execute it, or its output differs from fp32. Runtime and product measurements are reported separately and never act as publication certification.

Never overwrite the existing model repositories.

## Sources and metadata

- TinyBERT: `JueZhang/lingua2_tinybert_meetingbank_only` at `8b4a18ddae06f94c3e022b01b3beb994f75ebe76`
- MobileBERT: `JueZhang/lingua2_mobilebert_meetingbank_only` at `af78299502cbbb512ce8a664ff4b7ce7e6ff7e0a`
- New repositories:
  - `atjsh/llmlingua-2-js-tinybert-meetingbank-onnx-v4`
  - `atjsh/llmlingua-2-js-mobilebert-meetingbank-onnx-v4`
- Use `license: unknown` and neutral provenance. Do not claim qualification, accuracy, backend support, commercial suitability, Apache, or CC licensing.
- Record source and converter revisions, exact commands, filenames, byte sizes, SHA-256 hashes, converter validation output, and observed runtime results.

## Converter-trust publication

Run the pinned official Optimum/ONNX Runtime conversion recipes for:

- `fp32`
- `fp16`
- `int8`
- `uint8`
- `q8`
- `q4`
- `q4f16`
- `bnb4`

Upload every expected artifact emitted by a converter command that exits successfully. Validation warnings, runtime failures, numerical drift, or product-output drift do not block upload. Partial output from a failed command is not an emitted artifact and must not be uploaded.

Do not generate O2, O3, O4, or experimental MHA variants in this pass. Stop an individual M4 conversion job if it projects beyond one hour; no cloud fallback is authorized.

Store conversion scripts, converter logs, raw measurements, Pareto inputs, selected defaults, checksums, and the final manifest in the model repositories rather than the npm package repository.

## Demo selection

Publication does not imply demo eligibility. A candidate enters a backend-specific demo frontier only if it loads, returns the expected logits shape, returns finite output, and repeats deterministically three times on that backend.

Build separate WebGPU and WASM frontiers. Compare eligible candidates by:

- quality cost: `1 - macro retained-set F1` against that repository's fp32 conversion over all nine E2E texts, both `mean` and `first`, and rates `0.00` through `1.00` in `0.01` steps
- warm complete-compression median after five warmups and twenty measured runs
- selected artifact download bytes
- peak browser renderer RSS above the pre-model baseline

Log-transform latency, size, and memory; min-max normalize all four costs; treat a zero-range metric as zero. Remove dominated candidates and select the remaining point with the smallest equal-weight Euclidean distance to the ideal. Resolve ties by higher retained-set F1, lower latency p95, smaller download, then lexical dtype.

Quality, speed, size, and memory remain reported product measurements, not publication gates.

## Completion

- [x] `@atjsh/llmlingua-2` 3.0.0 is published.
- [ ] Both new repositories exist without changing legacy repositories.
- [ ] Every successfully emitted artifact and its provenance metadata are uploaded.
- [ ] Uploaded Hub files are verified against local SHA-256 values.
- [ ] Static WebGPU/WASM demo defaults are recorded from the deterministic Pareto calculation.
- [ ] Example build and browser smoke tests pass for selected defaults.
- [ ] Close this issue after publication and verification.

Related to #17 and #24. #31 remains a separate training task.

**Comment by atjsh** [comment]
### Inventory checkpoint

No source-model download or conversion has started yet.

- Release prerequisite satisfied: npm `latest` is `@atjsh/llmlingua-2@3.0.0`; Git tag `3.0.0` targets `70f5ff0442c43fa3119f57e2463dab7fde40a43b`.
- Package checkout is clean at that commit and will remain untouched.
- Hugging Face CLI is authenticated as `atjsh`. Legacy model repositories will not be modified.
- Free disk: about **18 GiB**. Cached reusable assets: Transformers.js v4 Node dependencies, Playwright Chromium, and the legacy TinyBERT fp32 ONNX (57,123,613 bytes) for comparison only. Neither pinned JueZhang source checkpoint is cached.
- Source-weight download is about **155 MB total** (TinyBERT ~57 MB; MobileBERT ~98 MB). The eight candidate variants are projected below **0.5 GB total**. Sequential conversion plus one isolated Python environment should peak below **4 GB**, leaving a >10 GiB reserve.
- Conversion implementation will follow the official `onnx-community/convert-to-onnx` Space at commit `af0ef3070cef5e863b371628ad3dcdc10de4d091`: Python 3.11, Torch 2.5.1, Optimum 2.1.0, optimum-onnx 0.1.0, Transformers 4.57.6, and the resolved ONNX/ONNX Runtime versions recorded in provenance. MHA/O3/O4 remain excluded.

Next: create the isolated tool environment, download each pinned source exactly once, export/quantize sequentially, then qualify before creating or uploading either target repository.

**Comment by atjsh** [comment]
### Qualification stopped at the fp32 prerequisite

The pinned source checkpoints were downloaded once and fp32 graphs were exported locally with the pinned official converter stack. Neither graph satisfies this issue's first correctness gate, so I stopped before quantization, Node/browser E2E, repository creation, or upload. The legacy repositories remain untouched.

| Model | PyTorch attention reference | ONNX | Max class-probability error | Gate | Result |
|---|---:|---:|---:|---:|---:|
| TinyBERT | default | opset 18 | `4.410743713378906e-6` | `<= 1e-6` | **FAIL** |
| TinyBERT | export-aligned `eager` | opset 18 | `2.0563602447509766e-6` | `<= 1e-6` | **FAIL** |
| MobileBERT | default | opset 18 | `3.635883331298828e-6` | `<= 1e-6` | **FAIL** |
| MobileBERT | export-aligned `eager` | opset 18 | `3.635883331298828e-6` | `<= 1e-6` | **FAIL** |

All nine test inputs were finite. The maxima occur on active, ordinary tokens (TinyBERT: `,`; MobileBERT: `wikipedia`), not padding or special-token positions. MobileBERT also triggered Optimum's own export warning: maximum logit difference `1.0` on its dummy validation input; across the fixed corpus the maximum logit difference was `2.75`. Its two-class softmax cancels most of that common logit offset, but not enough to pass the probability gate.

As a bounded cross-check, official opset 14 exports also failed (`2.0563602447509766e-6` TinyBERT; `3.635883331298828e-6` MobileBERT with export-aligned eager PyTorch). A PyTorch 2.7 reproduction matching the legacy TinyBERT graph's producer generation was worse, so the pinned toolchain was restored unchanged.

#### Reproducibility

- Source TinyBERT: `JueZhang/lingua2_tinybert_meetingbank_only@8b4a18ddae06f94c3e022b01b3beb994f75ebe76`
  - `model.safetensors`: `570887f3c4f4d3b614ced54536ed92ecc2fb423240dc73bd5c0fcb16e4592ee4` (57,020,992 bytes)
  - `config.json`: `1876b8870120061b6fcf2f99e8bff4c000b0d5e7f554b4dac14b647ce17cb550`
- Source MobileBERT: `JueZhang/lingua2_mobilebert_meetingbank_only@af78299502cbbb512ce8a664ff4b7ce7e6ff7e0a`
  - `model.safetensors`: `7c6e38183447c5d7846a10fbac680c942ee852584f95e9a0ad3077bbb1257849` (98,470,112 bytes)
  - `config.json`: `8290bfd1b650d3e858509b6503f27bf67d8e370a3a7a69594ce16cf3808ea97c`
- Input corpus: `src/e2e/long-texts.ts` at package commit `70f5ff0442c43fa3119f57e2463dab7fde40a43b`; file SHA-256 `481ac6ee205374aa50272c20388496fd051224c2f17da5436edd0566876744a3`.
- Converter reference: `onnx-community/convert-to-onnx@af0ef3070cef5e863b371628ad3dcdc10de4d091`.
- Resolved stack: Python 3.11.14; Torch 2.5.1; Transformers 4.57.6; Optimum 2.1.0; optimum-onnx 0.1.0; huggingface-hub 0.35.3; ONNX 1.22.0; ONNX Runtime 1.28.0; onnx-ir 0.2.1; onnxconverter-common 1.16.0.
- Export command (run once per pinned local source):

  ```sh
  optimum-cli export onnx --model <pinned-local-source> --task token-classification <output-directory>
  ```

| Local quarantined graph | Opset | Bytes | SHA-256 |
|---|---:|---:|---|
| TinyBERT fp32 | 18 | 57,089,382 | `1b908113de6195f4135b523011408e3148c518924788bc4af91a5a8c02ad80ec` |
| MobileBERT fp32 | 18 | 98,960,760 | `8902795483843cd2645ec5842980be4371a341bd774b5ec45f0804622b010d38` |

The comparison used ORT `CPUExecutionProvider`, tokenizer truncation at 512 tokens, the attention mask, full last-axis two-class softmax, and the maximum absolute class-probability difference over every active token in all nine inputs. Repeated inference was deterministic and finite.

#### Publication state

- `atjsh/llmlingua-2-js-tinybert-meetingbank-onnx-v4`: **not created**
- `atjsh/llmlingua-2-js-mobilebert-meetingbank-onnx-v4`: **not created**
- Hugging Face uploads: **none**
- Package/default changes: **none**; package worktree remains clean

Per the issue's “upload only qualified variants” requirement, lowering or waiving the fp32 gate would require an explicit issue decision. Until then, all candidate publication is blocked rather than mislabeled as qualified.

**Comment by atjsh** [comment]
The issue body now records the approved converter-trust policy. This supersedes the custom numerical and product-output publication gates discussed in earlier comments. Converter warnings, runtime failures, and output drift will be preserved as measurements in the neutral archive; they will not prevent upload when the pinned official converter command succeeds and emits the expected artifact. Demo eligibility and Pareto selection remain a separate product decision.

**Comment by atjsh** [comment]
### Converter-trust publication complete

The two new neutral conversion archives are now published:

- [TinyBERT ONNX v4](https://huggingface.co/atjsh/llmlingua-2-js-tinybert-meetingbank-onnx-v4) at `914542f15dfe5b24b8190f7f921cdf5254008647`
- [MobileBERT ONNX v4](https://huggingface.co/atjsh/llmlingua-2-js-mobilebert-meetingbank-onnx-v4) at `5bd77cc8a968650b2e5fb6bffaf75f66759feb0f`

For each source, the pinned official exporter emitted fp32 and the pinned
official converter emitted fp16, int8, uint8, q8 (`model_quantized.onnx`),
q4, q4f16, and bnb4. Every command exited successfully and all eight files
were uploaded. No O2/O3/O4 or experimental MHA graph was generated.

Each repository contains the exact source/converter revisions, source hashes,
official converter source, resolved environment, executed commands, complete
export/quantization logs, per-file byte sizes and SHA-256 hashes, a final file
manifest, and observed CPU runtime results. Jue-derived licensing metadata is
`unknown` and provenance is neutral.

Non-gating observation: ONNX Runtime 1.28.0 CPU loaded fp32, int8, uint8, q8,
q4, and bnb4 with finite `[1, 14, 2]` logits and three exact repetitions for
both models. The official fp16 and q4f16 outputs failed ORT's tensor-type load
validation; those successfully emitted artifacts remain published under the
converter-trust policy. MobileBERT's official fp32 export-difference warning
is preserved verbatim in its export log.

Verification downloaded both Hub snapshots at the exact commits and compared
all 26 uploaded files per repository with the local archive: no missing files,
unexpected files, or SHA-256 mismatches. The Hub-managed `.gitattributes` was
the sole additional file. Legacy repositories remain unchanged:

- TinyBERT: `a9af82841d3f815c9c492b13791c6517154791d3`
- MobileBERT: `900ed52628d7b153a276a220483d26d5f8dfe0f7`

Backend-specific Pareto measurement and demo selection continue separately;
runtime observations did not control publication.

**Comment by atjsh** [comment]
Closing on converter-trust publication completion. Both archive commit SHAs, provenance, runtime observations, byte-for-byte Hub verification, and unchanged legacy SHAs are recorded above. Pareto demo measurement continues independently; #31 was not modified.

**Comment by atjsh** [comment]
### Recorded demo Pareto outcome

The TinyBERT archive now includes the sanitized measurement harness, raw
backend-specific inputs, exact selector formula, static demo selections, and
file manifests at Hub commit
[`b6867947b3b5f4ad8c7e9239e8e1d8114322f35b`](https://huggingface.co/atjsh/llmlingua-2-js-tinybert-meetingbank-onnx-v4/commit/b6867947b3b5f4ad8c7e9239e8e1d8114322f35b).

The formula-produced static demo selections are:

- WebGPU: `bnb4` (ideal distance `0.48037161548137003`)
- WASM: `uint8` (ideal distance `0.47608842660279066`)

These are product-selection records, not qualification, recommendation,
commercial-suitability, or backend-support claims. Publication remained based
only on successful pinned official-converter emission.

Verification downloaded the exact final Hub snapshot and compared all 67
archive files with local bytes: no missing, unexpected, or SHA-256-mismatched
files. The Hub-managed `.gitattributes` is the sole additional file. The legacy
TinyBERT repository remains unchanged at
`a9af82841d3f815c9c492b13791c6517154791d3`.

**Comment by atjsh** [comment]
### TinyBERT archive audit follow-up

The final TinyBERT archive revision is now `96a5cab43d18b5723f29f797a1f0effc361a7034`, superseding the prior metrics revision without rewriting Hub history.

- Replaced one machine-specific shared-lock path in a raw operator-interruption diagnostic with the documented `$PARETO_LOCK_DIR` input.
- Recursively rescanned the exact downloaded revision, including decompressed raw metric archives, for `/Users`, `/private/tmp`, `/tmp`, and token/private-key patterns. The Pareto measurement tree contains no machine-specific absolute path or detected secret.
- The only remaining `/tmp` strings in the overall repository are preserved original converter commands, converter warning logs, and CPU runtime error messages; they are part of the exact conversion provenance and contain no credentials.
- All 67 local archive files match the exact downloaded Hub revision byte-for-byte, all 65 manifest checksums pass, and the legacy TinyBERT repository remains at `a9af82841d3f815c9c492b13791c6517154791d3`.

The recorded demo selection remains `{ webgpu: "bnb4", wasm: "uint8" }`.

**Comment by atjsh** [comment]
### TinyBERT source-provenance follow-up

The final TinyBERT archive revision is now `e49e9637ec1e9a88defd52b6422cf2d40e96d539`, superseding `96a5cab43d18b5723f29f797a1f0effc361a7034` without rewriting Hub history.

- Added the pinned upstream `state_dict.pth` to `provenance/source.json`: 57,039,714 bytes, SHA-256 `0d04a2f2d96c9ceb811590d03cb0445ea33fc856e53810c3f680674b361d6c4d`.
- Disclosed the source-directory staging step and that its exact shell spelling was not retained, rather than inventing a command. The export selected the recorded `model.safetensors`; `state_dict.pth` was not used.
- Verified the source inventory has exactly the same eight paths, sizes, and LFS hashes as `JueZhang/lingua2_tinybert_meetingbank_only@8b4a18ddae06f94c3e022b01b3beb994f75ebe76`.
- Used a parent-commit-guarded, five-path-only Hub commit. The latest repository still has the same 68 paths; every non-updated path has the same Hub fingerprint, the five uploaded files match locally byte-for-byte, and all 65 archive checksums pass.
- The recorded demo tuple remains `{ webgpu: "bnb4", wasm: "uint8" }`, and the legacy TinyBERT repository remains at `a9af82841d3f815c9c492b13791c6517154791d3`.

The equivalent MobileBERT upstream inventory/staging disclosure is already corrected locally and will be included in its guarded final measurement commit.

**Comment by atjsh** [comment]
MobileBERT converter-trust archive and the recorded demo Pareto measurements are now complete.

- Archive: [`atjsh/llmlingua-2-js-mobilebert-meetingbank-onnx-v4@3e0b45f7bafc20bb5f83a288ff02decf3facf900`](https://huggingface.co/atjsh/llmlingua-2-js-mobilebert-meetingbank-onnx-v4/tree/3e0b45f7bafc20bb5f83a288ff02decf3facf900)
- Source: `JueZhang/lingua2_mobilebert_meetingbank_only@af78299502cbbb512ce8a664ff4b7ce7e6ff7e0a`
- Official converter: `onnx-community/convert-to-onnx@af0ef3070cef5e863b371628ad3dcdc10de4d091`
- Emitted and uploaded after successful converter commands: `fp32`, `fp16`, `int8`, `uint8`, `q8`, `q4`, `q4f16`, and `bnb4`
- Recorded browser eligibility: `int8`, `uint8`, `q8`, `q4`, and `bnb4` on both WASM and physical WebGPU. The fp32 browser run crashed; fp16 and q4f16 reported the recorded ONNX type mismatch. These observations did not gate archive publication.
- Recorded static demo selection from the specified equal-weight Pareto formula: `{ webgpu: "bnb4", wasm: "bnb4" }`. This is a product-selection record, not a qualification, recommendation, commercial-suitability statement, or backend-support claim.

The repository contains the exact conversion scripts/logs, source and converter revisions, upstream source inventory (including the unused `state_dict.pth`), artifact byte sizes and SHA-256 hashes, raw deterministic-gzip measurements, sanitized browser/Node harness, selector inputs/output, and final manifest. The final Hub audit verified all 63 remote paths: 19 LFS objects by size/SHA-256 and 43 other archive files by downloaded SHA-256, plus `.gitattributes`. The legacy MobileBERT repository remains at `900ed52628d7b153a276a220483d26d5f8dfe0f7`.

#31 remains open and unchanged; no training work was performed.
